File: data/code_pairs/fortran/360.f90
module work_type_mod
    implicit none
    private

    public :: work_type, init_work_type

    type work_type
        integer :: ntotal_kernel
        integer :: ndimensions
        integer :: nvertices
        real, allocatable :: vertices(:,:)
        real, allocatable :: kernel_values(:,:)
    end type work_type

contains

    subroutine init_work_type(wt, nkern, ndim, nverts)
        type(work_type), intent(out) :: wt
        integer, intent(in) :: nkern, ndim, nverts

        wt%ntotal_kernel = nkern
        wt%ndimensions = ndim
        wt%nvertices = nverts

        allocate(wt%vertices(1:wt%ndimensions, 1:wt%nvertices))
        allocate(wt%kernel_values(1:wt%ntotal_kernel, 1:wt%nvertices))
        wt%vertices = 0.0
        wt%kernel_values = 0.0
    end subroutine init_work_type

end module work_type_mod

File: data/code_pairs/cpp/360.cpp
#ifndef WORKTYPE_H
#define WORKTYPE_H

#include <vector>

class WorkType {
public:
    int ntotal_kernel;
    int ndimensions;
    int nvertices;
    std::vector<std::vector<double>> vertices;
    std::vector<std::vector<double>> kernel_values;

    WorkType() : ntotal_kernel(0), ndimensions(0), nvertices(0) {}

    void init(int nkern, int ndim, int nverts) {
        ntotal_kernel = nkern;
        nvertices = nverts;
        ndimensions = ndim;

        vertices.resize(ndimensions, std::vector<double>(nvertices, 0.0));
        kernel_values.resize(ntotal_kernel, std::vector<double>(nvertices, 0.0));
    }
};

#endif // WORKTYPE_H